core/deployment/src/main/java/io/quarkus/deployment/builditem/ApplicationArchivesBuildItem.java

```diff
@@ -20,6 +20,11 @@ public final class ApplicationArchivesBuildItem extends SimpleBuildItem {
     public ApplicationArchivesBuildItem(ApplicationArchive root, Collection<ApplicationArchive> applicationArchives) {
         this.root = root;
         this.applicationArchives = applicationArchives;
+
+        HashSet<ApplicationArchive> ret = new HashSet<>(applicationArchives);
+        ret.add(root);
+
+        this.allArchives = Collections.unmodifiableSet(ret);
```

Comment:
You still need to make the `allArchives` field final.

Reply:
You are right. Done.